modules/auth/sso/sspi_windows.go

```
@@ -0,0 +1,307 @@
+	return models.IsSSPIEnabled()
+}
+
+// Priority determines the order in which authentication methods are executed.
+// The lower the priority, the sooner the plugin is executed.
+// The SSPI plugin should be executed last as it returns 401 status code
+// if negotiation fails or should continue, which would prevent other
+// authentication methods to execute at all.
+func (s *SSPI) Priority() int {
+	return 50000
+}
+
+// VerifyAuthData uses SSPI (Windows implementation of SPNEGO) to authenticate the request.
+// If authentication is successful, returs the corresponding user object.
+// If negotiation should continue or authentication fails, immediately returns a 401 HTTP
+// response code, as required by the SPNEGO protocol.
+func (s *SSPI) VerifyAuthData(ctx *macaron.Context, sess session.Store) *models.User {
+	if !s.shouldAuthenticate(ctx) {
+		return nil
+	}
+
+	cfg, err := s.getConfig()
+	if err != nil {
+		log.Error("could not get SSPI config: %v", err)
+		return nil
+	}
+
+	userInfo, outToken, err := sspiAuth.Authenticate(ctx.Req.Request, ctx.Resp)
+	if err != nil {
+		log.Warn("Authentication failed with error: %v\n", err)
+		sspiAuth.AppendAuthenticateHeader(ctx.Resp, outToken)
+
+		// Include the user login page in the 401 response to allow the user
+		// to login with another authentication method if SSPI authentication
+		// fails
+		addFlashErr(ctx, ctx.Tr("auth.sspi_auth_failed"))
+		ctx.Data["EnableOpenIDSignIn"] = setting.Service.EnableOpenIDSignIn
+		ctx.Data["EnableSSPI"] = true
+		ctx.HTML(401, string(tplSignIn))
+		return nil
+	}
+	if outToken != "" {
+		sspiAuth.AppendAuthenticateHeader(ctx.Resp, outToken)
+	}
+
+	username := sanitizeUsername(userInfo.Username, cfg)
+	log.Info("Authenticated as %s\n", username)
+	if len(username) == 0 {
+		return nil
+	}
+
+	user, err := models.GetUserByName(username)
+	if err != nil {
+		if !models.IsErrUserNotExist(err) {
+			log.Error("GetUserByName: %v", err)
+			return nil
```

This seems like an error we should return instead of nil.

---

Yes, will change the definition of `VerifyAuthData()` to also return an error:
`VerifyAuthData(ctx *macaron.Context, sess session.Store) (*models.User, error)`

But I am not sure how to handle that in `SignedInUser()`
https://github.com/go-gitea/gitea/blob/9f67df8cd75e7bceb3ee8077ccc4c84b455dfdf6/modules/auth/auth.go#L34-L52

 or `Contexter()`

https://github.com/go-gitea/gitea/blob/9f67df8cd75e7bceb3ee8077ccc4c84b455dfdf6/modules/context/context.go#L302-L304

Should those types of errors only be logged, just higher in the stack?

---

No, you're right. I see now that this kind of error is not returned to the user. It may be to avoid exposing anything sensitive to the client.

---

Ok, will leave the definition of `VerifyAuthData()` as it is, and will only log this error.